### PR TITLE
Add support for Emscripten and Visual Studio

### PR DESCRIPTION
**Pull request by dstaley**
This PR adds support for Emscripten (WebAssembly) and compiling using Visual Studio.

Both of these were achieved with the help of CMake, which has replaced the original use of Make. This makes certain things like library management a bit easier across platforms, and provides native Visual Studio integration without adding Visual Studio specific files.

The code was also lightly edited to support Visual Studio's C compiler, which required a few changes:

- Conditional wrapping of things that are available in gcc, but not MVC like `__builtin_unreachable` and `__builtin_expect`
- Using `uintptr_t` instead of `void*`
- Some changes to constant folding since MVC doesn't support it

Additionally, the main loop of the application was lifted to its own function to support Emscripten, which requires passing in the function to `emscripten_set_main_loop_arg`. Arguments for the loop function have been moved to a `render_context_t` type which just contains pointers to the values required to run the loop.

### CODE REVIEW

**Comment by UnrealPowerz** [comment]
Emscripten, CMake AND Visual Studio support -- love it! ❤️ Thank you for the contribution!

Looks great, just a few minor comments.

File: static/app.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import Factory from "./powar.js";
+
+const $i = document.getElementById;
+const $q = document.querySelector;
```
Comment (UnrealPowerz) [inline static/app.js]:
```suggestion
const $i = document.getElementById.bind(document);
const $q = document.querySelector.bind(document);
```

File: main.c [hunk 2]
```diff
@@ -3204,6 +3220,52 @@ static void tmrw_update(pw_context_t *ctx, int states) {
...
+    int ms_to_sleep = EXEC_BATCH_MS - (int)(seconds_elapsed * 1000);
+
+    // printf("Batch complete: %d states in %.6f s; sleeping for %d ms\n", (*(*context).ctx).states, seconds_elapsed, ms_to_sleep);
+
+    if (ms_to_sleep > 0) {
+        SDL_Delay(ms_to_sleep);
+    }
+#endif // !__EMSCRIPTEN__
+
+    (*(*context).ctx).states = 0;
```
Comment (UnrealPowerz) [inline main.c]:
```suggestion
    while (context->ctx->states < STATES_PER_BATCH) {
        int old_states = context->ctx->states;
        pw_step(context->ctx);
        tmrw_update(context->ctx, context->ctx->states - old_states);
        (*(context->count))++;
    }
    //rtc_update(&ctx.rtc);
    //int old_keys = ctx.keys_pressed;
    context->ctx->keys_pressed = sdl_poll(context->ctx->keys_pressed, context->should_redraw);
    // TODO: maybe invert keys pressed ?
    portb_update(&context->ctx->portb, context->ctx->keys_pressed);
    if (*(context->should_redraw)) {
        sdl_draw(&context->ctx->lcd);
        *(context->should_redraw) = 0;
    }

#ifndef __EMSCRIPTEN__
    Uint32 end = SDL_GetPerformanceCounter();
    float seconds_elapsed = (end - start) / (float)SDL_GetPerformanceFrequency();
    int ms_to_sleep = EXEC_BATCH_MS - (int)(seconds_elapsed * 1000);

    // printf("Batch complete: %d states in %.6f s; sleeping for %d ms\n", (*(*context).ctx).states, seconds_elapsed, ms_to_sleep);

    if (ms_to_sleep > 0) {
        SDL_Delay(ms_to_sleep);
    }
#endif // !__EMSCRIPTEN__

    context->ctx->states = 0;
```

**Review by UnrealPowerz — CHANGES_REQUESTED** [review]
My apologies for the `main.c` file BTW, I should really split it out and clean it up one of these days.

Great work!

**Comment by dstaley** [comment]
Thanks for the review! I didn't know about the right arrow notation, but that makes that function read _so_ much better, so thank you!

**Pull request merged**